File: medium/add-two-numbers.py
#coding
"""
You are given two non-empty linked lists representing
two non-negative integers. The digits are stored in
reverse order and each of their nodes contain a single
digit. Add the two numbers and return it as a linked list.

You may assume the two numbers do not contain any
leading zero, except the number 0 itself.

Input: (2 -> 4 -> 3) + (5 -> 6 -> 4)
Output: 7 -> 0 -> 8
"""
# Definition for singly-linked list.
class ListNode(object):
    def __init__(self, x, n=None):
        self.val = x
        self.next = n

class Solution(object):
    def addTwoNumbers(self, l1, l2):
        """
        :type l1: ListNode
        :type l2: ListNode
        :rtype: ListNode
        """
        higher = 0
        result = root = ListNode(0)
        while l1 or l2 or higher:
            num1 = num2 = 0
            if l1:
                num1 = l1.val
                l1 = l1.next
            if l2:
                num2 = l2.val
                l2 = l2.next
            higher, value = divmod(num1 + num2 + higher, 10)
            root.next = ListNode(value)
            root = root.next
        return result.next


if __name__ == "__main__":
    l1 = ListNode(2, ListNode(4, ListNode(3)))
    l2 = ListNode(5, ListNode(6, ListNode(4)))
    sol = Solution()
    result = sol.addTwoNumbers(l1, l2)
    while result:
        print(result.val)
        result = result.next

File: medium/add-two-numbers.go
package main

import "fmt"

type ListNode struct {
	Val  int
	Next *ListNode
}

/**
https://leetcode.com/problems/add-two-numbers/
 * Definition for singly-linked list.
 * type ListNode struct {
 *     Val int
 *     Next *ListNode
 * }
*/
func addTwoNumbers(l1 *ListNode, l2 *ListNode) *ListNode {
	result := &ListNode{}
	current := result

	higher := 0
	for l1 != nil || l2 != nil || higher > 0 {
		a, b := 0, 0
		if l1 != nil {
			a = l1.Val
			l1 = l1.Next
		}
		if l2 != nil {
			b = l2.Val
			l2 = l2.Next
		}
		sum := a + b + higher
		current.Next = &ListNode{
			Val: sum % 10,
		}
		higher = sum / 10
		current = current.Next
	}
	return result.Next
}

func listToListNode(l []int) *ListNode {
	ln := &ListNode{}
	var current *ListNode = ln
	for _, i := range l {
		current.Next = &ListNode{
			Val: i,
		}
		current = current.Next
	}
	return ln.Next
}

func printNode(l *ListNode) []int {
	a := []int{}
	for l != nil {
		a = append(a, l.Val)
		l = l.Next
	}
	return a
}

func main() {
	// l1 := ListNode{
	// 	Val: 2,
	// 	Next: &ListNode{
	// 		Val: 4,
	// 		Next: &ListNode{
	// 			Val: 3,
	// 		},
	// 	},
	// }
	// l2 := ListNode{
	// 	Val: 5,
	// 	Next: &ListNode{
	// 		Val: 6,
	// 		Next: &ListNode{
	// 			Val: 4,
	// 		},
	// 	},
	// }
	// fmt.Println(addTwoNumbers(&l1, &l2))

	fmt.Println(printNode(addTwoNumbers(listToListNode([]int{9, 9, 9, 9, 9, 9, 9}), listToListNode([]int{9, 9, 9, 9}))))
	// fmt.Println(printNode(addTwoNumbers(listToListNode([]int{0}), listToListNode([]int{0}))))

}
